controllers/team_controller.go

```diff
@@ -165,7 +199,69 @@ func (t *TeamReconciler) checkMetricNSForTeamIsDeleted(ctx context.Context, req
 
 // SetupWithManager sets up the controller with the Manager.
 func (t *TeamReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	labelPredicate := predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		labels := obj.GetLabels()
+		_, exists := labels["snappcloud.io/team"]
+		return exists
+	})
+
+	mapFunc := func(a client.Object) []reconcile.Request {
+		ctx := context.Background()
+		log := log.FromContext(ctx)
+
+		var requests []reconcile.Request
+
+		var teamList teamv1alpha1.TeamList
+		if err := mgr.GetClient().List(ctx, &teamList, &client.ListOptions{}); err != nil {
+			// Handle error
+			log.Error(err, "Unable to list team resources")
+			return nil
+		}
+
+		for _, team := range teamList.Items {
+			requests = append(requests, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      team.Name,
+					Namespace: team.Namespace,
+				},
+			})
+		}
+
+		return requests
+	}
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&teamv1alpha1.Team{}).
+		Watches(
+			&source.Kind{Type: &corev1.Namespace{}},
+			handler.EnqueueRequestsFromMapFunc(mapFunc),
+			builder.WithPredicates(labelPredicate),
+		).
 		Complete(t)
 }
+
+func (t *TeamReconciler) finalizeNamespace(ctx context.Context, req ctrl.Request, ns *corev1.Namespace, team *teamv1alpha1.Team) error {
+
+	for i, namespace := range team.Spec.Namespaces {
+		if namespace == ns.Name {
+			team.Spec.Namespaces = append(team.Spec.Namespaces[:i], team.Spec.Namespaces[i+1:]...)
+			break
+		}
+	}
+	// Remove the 'snappcloud.io/team' label from the namespace
+	_, ok := ns.Labels["snappcloud.io/team"]
+	if ok {
+		delete(ns.Labels, "snappcloud.io/team")
+	}
+	if err := t.Client.Update(ctx, ns); err != nil {
+		return err
+	}
```

Comment:
These three lines can be moved inside the `if` block above. The update might be a redundant request if the `namespace` is not changed.